astropy/modeling/fitting.py

```diff
@@ -446,8 +461,18 @@ def __call__(self, model, x, y, z=None, weights=None, rcond=None):
                     # but z has a second axis for the model set. NB. This is
                     # ~5-10x faster than using rollaxis.
                     rhs = z.T if model_axis == 0 else z
+
+                if weights is not None:
+                    # Same for weights
+                    if weights.ndim > 2:
+                        weights = np.rollaxis(z, model_axis, z.ndim)
+                        weights = weights.reshape(-1, weights.shape[-1])
+                    else:
+                        weights = weights.flatten()
```

Comment:
This I'm not really sure about, because it doesn't match the else clause for the data just above, but I didn't really understand this case.

Reply:
Yes, that's one of the main things I didn't understand. It doesn't look right but I'll try to figure it out.

Reply:
I believe I've figured out what to do about this, but haven't done a lot more after a somewhat interrupted day (mostly by screaming kids, which is a bit of a nightmare when trying to keep all these different cases in one's head...). The current `else` here is handling multiple cases that need separating and there are one or two other changes needed to fix it, including an additional test or two for a case that currently fails.